src/encord_active/analysis/executor.py

```diff
@@ -0,0 +1,1177 @@
+                    )
+                self._execute_stage_1_metrics_no_batching(
+                    prev_frame=None,
+                    current_frame=frame_input,
+                    next_frame=None,
+                    current_frame_gt=gt_input,
+                    du_hash=du_meta_job.du_hash,
+                    frame=du_meta_job.frame,
+                )
+
+        # Thread-pool execute by chunks when running on cpu mode
+        if self.config.device.type == "cpu":
+            with ThreadPoolExecutor(max_workers=self.pool_size) as pool:
+                tqdm_desc = tqdm(total=len(images), desc="Metric Computation Stage 1: Images")
+                for i in range(0, len(images), 100):
+                    images_group = images[i : i + 100]
+                    results = []
+                    for data_metadata, du_meta in images_group:
+                        results.append(pool.submit(_job, data_meta_job=data_metadata, du_meta_job=du_meta))
+                    for result in results:
+                        # Wait for completion.
+                        result.result(timeout=3000.0)
+                        tqdm_desc.update(1)
+        else:
+            for data_metadata, du_meta in tqdm(images, desc="Metric Computation Stage 1: Images"):
+                _job(data_meta_job=data_metadata, du_meta_job=du_meta)
+
+    def _stage_1_image_batching(
+        self,
+        values: List[Tuple[ProjectDataMetadata, ProjectDataUnitMetadata]],
+        database_dir: Path,
+        project_hash: uuid.UUID,
+        project_ssh_path: Optional[str],
+        batching: int,
+        gt_lookup: Optional[Dict[Tuple[uuid.UUID, int], ProjectDataUnitMetadata]],
+    ) -> None:
+        grouped_values: Dict[Tuple[int, int], List[Tuple[ProjectDataMetadata, ProjectDataUnitMetadata]]] = {}
+        for data_metadata, du_metadata in values:
+            grouped_values.setdefault((du_metadata.width, du_metadata.height), []).append((data_metadata, du_metadata))
+        for group in grouped_values.values():
+            for i in range(0, len(group), batching):
+                grouped_batch = group[i : i + batching]
+                frame_inputs = [
+                    self._get_frame_input(
+                        image=self._open_image(
+                            uri=du_meta.data_uri,
+                            database_dir=database_dir,
+                            project_hash=project_hash,
+                            project_ssh_path=project_ssh_path,
+                            label_hash=meta.label_hash,
+                            data_hash=du_meta.data_hash,
+                            du_hash=du_meta.du_hash,
+                        ),
+                        du_meta=du_meta,
+                    )
+                    for meta, du_meta in grouped_batch
+                ]
+
+                batch_inputs = BaseFrameBatchInput(
+                    images=torch.stack([frame.image for frame in frame_inputs]),
```

Comment:
Images may have different sizes, so the stack operation here fails in that case right?